File: 백준/Gold/7569. 토마토/토마토.java
import java.util.*;
import java.io.*;

public class Main {
    public static StringBuilder sb = new StringBuilder();
    public static StringTokenizer st;
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    int garo,sero,h;
    int[][][] map;
    boolean[][][] visited;
    int[][][] dist;
    Queue<int[]> q = new LinkedList<>();
    int[] dx = {-1, 0, 1, 0, 0, 0};
    int[] dy = {0, 1, 0, -1, 0, 0};
    int[] dz = {0, 0, 0, 0, 1, -1};

    public void sol() throws Exception {
        String[] split = br.readLine().split(" ");
        garo = Integer.parseInt(split[0]);
        sero = Integer.parseInt(split[1]);
        h = Integer.parseInt(split[2]);
        map = new int[garo+1][sero+1][h+1];
        visited = new boolean[garo+1][sero+1][h+1];
        dist = new int[garo+1][sero+1][h+1];

        for(int i=1;i<=h;i++){
            for(int j=1;j<=sero;j++){
                String[] input = br.readLine().split(" ");
                for(int k=0;k< garo;k++){
                    map[k+1][j][i] = Integer.parseInt(input[k]);
                }
            }
        }

        for(int i = 1; i <= h; i++){
            for(int j = 1; j <= sero; j++){
                for(int k = 1; k <= garo; k++){
                    if (map[k][j][i] == 1 && visited[k][j][i] == false) {
                        q.add(new int[]{k, j, i});
                    }
                }
            }
        }
        bfs();
        if (hasZero()) { // 0이 있다면 안익은 토마토 있음
            System.out.println("-1");
        } else {
            int max = max();
            System.out.println(max-1);
        }
        br.close();
    }

    public boolean hasZero() {
        for(int i = 1; i <= h; i++){
            for(int j = 1; j <= sero; j++){
                for(int k = 1; k <= garo; k++){
                    if(map[k][j][i]==0)
                        return true;
                }
            }
        }
        return false;
    }

    public int max(){
        int max = 0;
        for(int i = 1; i <= h; i++){
            for(int j = 1; j <= sero; j++){
                for(int k = 1; k <= garo; k++){
                    if (map[k][j][i] > max)
                        max = map[k][j][i];
                }
            }
        }
        return max;
    }

    public void bfs() {
        while (!q.isEmpty()) {
            int[] tmp = q.poll();
            int x = tmp[0];
            int y = tmp[1];
            int z = tmp[2];
            for (int k = 0; k < 6; k++) {
                int nx = x + dx[k];
                int ny = y + dy[k];
                int nz = z + dz[k];
                if(nx>0 && ny>0 && nz>0 && nx<=garo && ny<=sero && nz<=h){
                    if(map[nx][ny][nz] == 0 && visited[nx][ny][nz] == false){
                        q.add(new int[]{nx, ny, nz});
                        visited[nx][ny][nz] = true;
                        map[nx][ny][nz] = map[x][y][z] + 1;
                    }
                }
            }
        }
    }

    public void show() {
        for(int i = 1; i <= h; i++){
            for(int j = 1; j <= sero; j++){
                for(int k = 1; k <= garo; k++){
                    System.out.print(map[k][j][i] + " ");
                }
                System.out.println();
            }
            System.out.println();
        }
    }

    public static void main(String[] args) throws Exception {
        new Main().sol();
    }
}

File: 백준/Gold/7569. 토마토/토마토.py
from collections import deque 
M, N, H = map(int, input().split())
box = []
visited = [[[0 for _ in range(M)] for _ in range(N)] for _ in range(H)]
dx = [0,1,0,-1,0,0]
dy = [1,0,-1,0,0,0]
dz = [0,0,0,0,-1,1]
q = deque()

for _ in range(H):
    sub = []
    for _ in range(N):
        sub.append(list(map(int, input().split())))
    box.append(sub)
def bfs():
    while q:
        h,n,m = q.popleft()
        for k in range(6):
            nn = n + dx[k]
            nm = m + dy[k]
            nh = h + dz[k]
            if nn >= 0 and nm >= 0 and nh >= 0 and nn < N and nm < M and nh < H:
                if box[nh][nn][nm] == 0 and visited[nh][nn][nm] == 0:
                    box[nh][nn][nm] = box[h][n][m] + 1
                    q.append((nh,nn,nm))
                    visited[nh][nn][nm] = 1

for height in range(H):
    for sero in range(N):
        for garo in range(M):
            if box[height][sero][garo] == 1 and visited[height][sero][garo] == 0:
                q.append((height,sero,garo))
                visited[height][sero][garo] = 1
bfs()
ans = 0
for a in box:
    for b in a:
        for c in b:
            if c == 0:
                print(-1)
                exit(0)
        ans = max(ans, max(b))
print(ans-1)
